File: ui/android_native/MainActivity.java
package com.dancorsoodessa.jarvis_ui;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.Locale;

import io.flutter.embedding.android.FlutterActivity;
import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodChannel;

/**
 * Small Android-native host layer. Flutter owns UI/state; Android owns the
 * platform speech recognizer. No Python process or executable is started.
 */
public class MainActivity extends FlutterActivity {
    private static final String VOICE_CHANNEL = "com.dancorsoodessa.jarvis/voice";
    private static final String VOICE_EVENTS = "com.dancorsoodessa.jarvis/voice_events";
    private static final int RECORD_AUDIO_REQUEST = 4107;

    private SpeechRecognizer recognizer;
    private EventChannel.EventSink voiceSink;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private boolean voiceRequested = false;

    @Override
    public void configureFlutterEngine(@NonNull FlutterEngine flutterEngine) {
        super.configureFlutterEngine(flutterEngine);

        new MethodChannel(flutterEngine.getDartExecutor().getBinaryMessenger(), VOICE_CHANNEL)
                .setMethodCallHandler((call, result) -> {
                    try {
                        switch (call.method) {
                            case "isAvailable":
                                result.success(SpeechRecognizer.isRecognitionAvailable(this));
                                break;
                            case "start":
                                startVoice(result);
                                break;
                            case "stop":
                                stopVoice();
                                result.success(null);
                                break;
                            default:
                                result.notImplemented();
                        }
                    } catch (Exception e) {
                        result.error("VOICE_HOST_ERROR", e.getMessage(), null);
                    }
                });

        new EventChannel(flutterEngine.getDartExecutor().getBinaryMessenger(), VOICE_EVENTS)
                .setStreamHandler(new EventChannel.StreamHandler() {
                    @Override
                    public void onListen(Object arguments, EventChannel.EventSink events) {
                        voiceSink = events;
                    }

                    @Override
                    public void onCancel(Object arguments) {
                        voiceSink = null;
                    }
                });
    }

    private void startVoice(MethodChannel.Result result) {
        if (!SpeechRecognizer.isRecognitionAvailable(this)) {
            result.error("VOICE_UNAVAILABLE", "Android speech recognition service is unavailable", null);
            return;
        }

        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            voiceRequested = true;
            requestPermissions(new String[]{Manifest.permission.RECORD_AUDIO}, RECORD_AUDIO_REQUEST);
            result.error("PERMISSION_REQUIRED", "Microphone permission is required", null);
            return;
        }

        ensureRecognizer();
        startListeningSafely();
        result.success(true);
    }

    private void ensureRecognizer() {
        if (recognizer != null) return;
        recognizer = SpeechRecognizer.createSpeechRecognizer(this);
        recognizer.setRecognitionListener(new RecognitionListener() {
            @Override public void onReadyForSpeech(Bundle params) { emit("status", "ready"); }
            @Override public void onBeginningOfSpeech() { emit("status", "speaking"); }
            @Override public void onRmsChanged(float rmsdB) { }
            @Override public void onBufferReceived(byte[] buffer) { }
            @Override public void onEndOfSpeech() { emit("status", "ended"); }

            @Override public void onError(int error) {
                emit("error", errorName(error));
            }

            @Override public void onResults(Bundle results) {
                ArrayList<String> matches = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                if (matches != null && !matches.isEmpty()) {
                    emit("result", matches.get(0));
                }
            }

            @Override public void onPartialResults(Bundle results) {
                ArrayList<String> matches = results.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
                if (matches != null && !matches.isEmpty()) {
                    emit("partial", matches.get(0));
                }
            }

            @Override public void onEvent(int eventType, Bundle params) { }
        });
    }

    private void startListeningSafely() {
        if (recognizer == null) return;
        try {
            recognizer.cancel();
            Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
            intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
            intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, "ru-RU");
            intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_PREFERENCE, "ru-RU");
            intent.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true);
            intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3);
            recognizer.startListening(intent);
        } catch (Exception e) {
            emit("error", "START_FAILED: " + e.getMessage());
        }
    }

    private void stopVoice() {
        voiceRequested = false;
        if (recognizer != null) {
            try { recognizer.cancel(); } catch (Exception ignored) { }
        }
    }

    private void emit(String type, String value) {
        if (voiceSink == null) return;
        mainHandler.post(() -> {
            if (voiceSink == null) return;
            java.util.HashMap<String, Object> event = new java.util.HashMap<>();
            event.put("type", type);
            event.put("value", value);
            voiceSink.success(event);
        });
    }

    private String errorName(int error) {
        switch (error) {
            case SpeechRecognizer.ERROR_AUDIO: return "ERROR_AUDIO";
            case SpeechRecognizer.ERROR_CLIENT: return "ERROR_CLIENT";
            case SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS: return "ERROR_INSUFFICIENT_PERMISSIONS";
            case SpeechRecognizer.ERROR_NETWORK: return "ERROR_NETWORK";
            case SpeechRecognizer.ERROR_NETWORK_TIMEOUT: return "ERROR_NETWORK_TIMEOUT";
            case SpeechRecognizer.ERROR_NO_MATCH: return "ERROR_NO_MATCH";
            case SpeechRecognizer.ERROR_RECOGNIZER_BUSY: return "ERROR_RECOGNIZER_BUSY";
            case SpeechRecognizer.ERROR_SERVER: return "ERROR_SERVER";
            case SpeechRecognizer.ERROR_SPEECH_TIMEOUT: return "ERROR_SPEECH_TIMEOUT";
            default: return "ERROR_" + error;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == RECORD_AUDIO_REQUEST) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                emit("permission", "granted");
                if (voiceRequested) {
                    voiceRequested = false;
                    ensureRecognizer();
                    startListeningSafely();
                }
            } else {
                emit("permission", "denied");
            }
        }
    }

    @Override
    protected void onDestroy() {
        stopVoice();
        if (recognizer != null) {
            try { recognizer.destroy(); } catch (Exception ignored) { }
            recognizer = null;
        }
        super.onDestroy();
    }
}

File: app/src/main/kotlin/com/dancorsoodessa/jarvis_ui/MainActivity.kt
package com.dancorsoodessa.jarvis_ui

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import android.speech.tts.TextToSpeech
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import org.json.JSONArray
import java.util.Locale

class MainActivity : FlutterActivity(), RecognitionListener {
    private val voiceChannel = "busya.voice"
    private val eventsChannel = "busya.voice.events"
    private val recordAudioRequest = 4101
    private val memoryPrefs = "busya_memory"
    private val memoryKey = "lessons"
    private var methodChannel: MethodChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var recognizer: SpeechRecognizer? = null
    private var tts: TextToSpeech? = null
    private var ttsReady = false
    private var pendingSpeech: String? = null
    private var disposed = false
    private var listening = false

    override fun configureFlutterEngine(flutterEngine: FlutterEngine) {
        super.configureFlutterEngine(flutterEngine)
        methodChannel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, voiceChannel)
        methodChannel?.setMethodCallHandler { call, result ->
            when (call.method) {
                "initialize" -> result.success(initializeVoice())
                "start" -> { startListening(); result.success(null) }
                "stop" -> { stopListening(); result.success(null) }
                "speak" -> { speak(call.argument<String>("text") ?: ""); result.success(null) }
                "load_learning" -> result.success(loadLearning())
                "save_learning" -> {
                    val items = call.argument<List<*>>("items")?.mapNotNull { it?.toString() } ?: emptyList()
                    saveLearning(items)
                    result.success(null)
                }
                "clear_learning" -> { clearLearning(); result.success(null) }
                "dispose" -> { releaseVoice(); result.success(null) }
                else -> result.notImplemented()
            }
        }
        EventChannel(flutterEngine.dartExecutor.binaryMessenger, eventsChannel)
            .setStreamHandler(object : EventChannel.StreamHandler {
                override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) { eventSink = sink }
                override fun onCancel(arguments: Any?) { eventSink = null }
            })
    }

    private fun initializeVoice(): Boolean {
        if (disposed) return false
        if (!SpeechRecognizer.isRecognitionAvailable(this)) {
            eventSink?.success("__ERROR__")
            return false
        }
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), recordAudioRequest)
            return false
        }
        ensureRecognizer()
        ensureTts()
        eventSink?.success("__READY__")
        return true
    }

    private fun ensureRecognizer() {
        if (recognizer != null || disposed) return
        recognizer = SpeechRecognizer.createSpeechRecognizer(this).also { it.setRecognitionListener(this) }
    }

    private fun ensureTts() {
        if (tts != null || disposed) return
        tts = TextToSpeech(this) { status ->
            ttsReady = status == TextToSpeech.SUCCESS
            if (!ttsReady) { eventSink?.success("__ERROR__"); return@TextToSpeech }
            val languageResult = tts?.setLanguage(Locale("ru", "RU"))
            tts?.setSpeechRate(0.48f)
            if (languageResult == TextToSpeech.LANG_MISSING_DATA || languageResult == TextToSpeech.LANG_NOT_SUPPORTED) {
                eventSink?.success("__ERROR__")
            } else {
                eventSink?.success("__TTS_READY__")
                val queued = pendingSpeech
                pendingSpeech = null
                if (!queued.isNullOrBlank()) speak(queued)
            }
        }
    }

    private fun startListening() {
        if (disposed || listening) return
        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.RECORD_AUDIO), recordAudioRequest)
            return
        }
        ensureRecognizer()
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
            putExtra(RecognizerIntent.EXTRA_LANGUAGE, "ru-RU")
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_PREFERENCE, "ru-RU")
            putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, false)
            putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3)
            putExtra(RecognizerIntent.EXTRA_SPEECH_INPUT_COMPLETE_SILENCE_LENGTH_MILLIS, 1200L)
            putExtra(RecognizerIntent.EXTRA_SPEECH_INPUT_MINIMUM_LENGTH_MILLIS, 250L)
        }
        try { listening = true; recognizer?.startListening(intent) }
        catch (exception: Exception) {
            listening = false
            eventSink?.success("__ERROR__")
        }
    }

    private fun stopListening() {
        listening = false
        try { recognizer?.cancel() } catch (_: Exception) { }
    }

    private fun speak(text: String) {
        val clean = text.trim()
        if (clean.isEmpty() || disposed) return
        ensureTts()
        if (!ttsReady) { pendingSpeech = clean; return }
        try { tts?.speak(clean, TextToSpeech.QUEUE_FLUSH, null, "busya_reply") }
        catch (_: Exception) { eventSink?.success("__ERROR__") }
    }

    private fun loadLearning(): List<String> {
        val raw = getSharedPreferences(memoryPrefs, MODE_PRIVATE).getString(memoryKey, "[]") ?: "[]"
        return try {
            val json = JSONArray(raw)
            (0 until json.length()).mapNotNull { index -> json.optString(index, null) }
        } catch (_: Exception) { emptyList() }
    }

    private fun saveLearning(items: List<String>) {
        val bounded = items.map { it.trim() }.filter { it.isNotEmpty() }.takeLast(100)
        val json = JSONArray()
        bounded.forEach { json.put(it.take(1000)) }
        getSharedPreferences(memoryPrefs, MODE_PRIVATE).edit().putString(memoryKey, json.toString()).apply()
    }

    private fun clearLearning() {
        getSharedPreferences(memoryPrefs, MODE_PRIVATE).edit().remove(memoryKey).apply()
    }

    private fun releaseVoice() {
        if (disposed) return
        disposed = true
        stopListening()
        try { recognizer?.destroy() } catch (_: Exception) { }
        recognizer = null
        try { tts?.stop(); tts?.shutdown() } catch (_: Exception) { }
        tts = null
        ttsReady = false
        pendingSpeech = null
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == recordAudioRequest) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ensureRecognizer(); ensureTts(); eventSink?.success("__READY__")
            } else eventSink?.success("__ERROR__")
        }
    }

    override fun onReadyForSpeech(params: Bundle?) = Unit
    override fun onBeginningOfSpeech() = Unit
    override fun onRmsChanged(rmsdB: Float) = Unit
    override fun onBufferReceived(buffer: ByteArray?) = Unit
    override fun onEndOfSpeech() { listening = false; eventSink?.success("__END__") }
    override fun onError(error: Int) { listening = false; eventSink?.success("__ERROR__") }
    override fun onResults(results: Bundle?) {
        listening = false
        val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
        val phrase = matches?.firstOrNull()?.trim().orEmpty()
        if (phrase.isNotEmpty()) eventSink?.success(phrase) else eventSink?.success("__END__")
    }
    override fun onPartialResults(partialResults: Bundle?) = Unit
    override fun onEvent(eventType: Int, params: Bundle?) = Unit

    override fun onDestroy() { releaseVoice(); super.onDestroy() }
}
